docs/proposals/default-service-plans.md

```diff
@@ -0,0 +1,544 @@
+
+
+#### Consumer
+
+End users should be able to install a chart on a cluster with any broker that supports this proposal, and not be required to specify any broker-specific values:
+
+
+``` console
+$ helm install azure/wordpress
+# Provision a new service of type mysql and inject it into Wordpress
+```
+
+
+If the cluster operator has created custom classes, that can be specified when installing the chart:
+
+
+``` console
+$ helm install azure/wordpress \
+    --set className=custom-mysql --set planName=ha-mysql
+```
+
+
+
+## Implementation
+
+A key goal of the implementation is to avoid going off and rewriting or requiring a long term branch or fork, etc. I'd like to see this developed using a feature flag, and making all new fields optional.
+
+
+### Service Class
+
+
+| Field | Description |
+|-------|-------------|
+| Service Type | Examples: mysql, redis|
+| <p>Default Provision Parameters</p><p>Default Bind Parameters</p> | <p>Default parameters to supply to the broker during provisioning and binding.</p> <p>OSB API V3 - Make the schema required, and defaults optional.|
+| Default Bind Response Transform | <p>Default json key mapping to transform a non-standard broker response into something that can be relied upon.</p> <p>For example, if a broker returns "database.name" (a nested value) in the response, it can be changed to "database" to un-nest it and apply a standard name.</p>|
+| Controller Reference |<p>Set by Service Catalog when populating the catalog from a broker list to indicate that it is managed by the controller.</p> <p>When not present, the class is ignored during a relist. This gives operators a way to define custom classes.</p>|
+
+Notes
+
+* Relist needs to take into account if the class is managed by the controller or was created by the cluster operator.
+* Defaults can be defined at the Cluster or Namespace level.
+* For simplification, only broker bind response bodies can be transformed, not provision, since that's where we require that functionality today.
+
+
+### Service Plan
+
+| Field | Description |
+|-------|-------------|
+| Service Type | Examples: mysql, redis |
+| annotation: is-default-plan=true | <p>Specifies whether or not a Service Plan is the default for its Service Type.</p> <p>When annotated as the default and a Service Type is set, this plan is used to dynamically provision an instance for that type.</p> |
+| Controller Reference|<p>Set by Service Catalog when populating the catalog from a broker list to indicate that it is managed by the controller.</p> <p>When not present, the plan is ignored during a relist. This gives operators a way to define custom plans.</p> |
+
+
+Notes:
+
+* The controller will enforce that only one default is set per service type.
+* For simplification, defaults are set at the class level only, plans cannot override them.
```

Comment:
I think this needs to be sorted out now. Perhaps both class and plan have defaults and there is a merge rule, but the defaults on class are applied and merged into the defaults for plan, and that result is the thing that is used for parameter defaults

Reply:
Agreed! 👍